src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py

```diff
@@ -159,12 +159,14 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return x
 
 
-def apply_rotary_pos_emb_flashatt(tensor: torch.Tensor, freqs: torch.Tensor) -> torch.Tensor:
-    tensor_ = tensor.float()
-    cos = freqs.cos()
-    sin = freqs.sin()
-    output = apply_rotary_emb(tensor_, cos, sin).type_as(tensor)
-    return output
+def apply_rotary_pos_emb_flashatt(
+    q: torch.Tensor, k: torch.Tensor, cos: torch.Tensor, sin: torch.Tensor
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    cos = cos.chunk(2, dim=-1)[0].contiguous()
+    sin = sin.chunk(2, dim=-1)[0].contiguous()
```

Comment:
why do we need chunk here?

Reply:
Because they use `apply_rotary_emb` provided by `flash-attn` when attn impl is FA2. It only needs half of the cos/sin embeddings, i.e. `[seq_len, head_dim//2]`, while the naive `apply_rotary_pos_emb_vision` for eager/sdpa requires full embeds (`[seq_len, head_dim]`). Conventionally we pass full embeddings to this API so I have to slice it for FA2. Do you have an elegant and efficient implementation for this?

Reply:
nope thanks for explaining!